urld73c47f3e5_lindenb_dataindexer_tgz-pJ8-test_github_lindenb_dataindexer_SimpleIntegerJ8/src/com/github/lindenb/dataindexer/AbstractDatabaseReader.java

```diff
@@ -0,0 +1,205 @@
+package com.github.lindenb.dataindexer;
+
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.AbstractList;
+import java.util.List;
+import java.util.logging.Logger;
+
+/**
+ * Abstract class  to random-access a list of item 'T' 
+ * @author lindenb
+ *
+ * @param <T> the type of item
+ * @param <CONFIG> the configuration type
+ */
+public class AbstractDatabaseReader<T,CONFIG extends AbstractConfig<T>>
+	implements Closeable
+	{
+	protected static final Logger LOG=Logger.getLogger("com.github.lindenb.dataindexer");
+	/** the configuration file */
+	private CONFIG config;
+	/** number of items (size) stored */
+	private long numberOfItems=0;
+	/** handle to the index file , if needed. A fixed-size data file doesn't need this */
+	private RandomAccessFile indexFile=null;
+	/** handle to the data-file */
+	private RandomAccessInput dataFile=null;
+	/** constructor from a configuration */
+	protected  AbstractDatabaseReader(CONFIG config) throws IOException
+		{
+		this.config=config;
+		}
+	
+	/** returns wether the datafile is currently opened */
+	public boolean isOpen()
+		{
+		return dataFile!=null;
+		}
+	
+	
+	protected void validateConfig() throws IOException
+		{
+		if(getConfig()==null) throw new IOException("config is null");
+		getConfig().validateForReading();
+		}
+	
+	/** open the datafile and , if needed, the index file .
+	 * Does nothing if the datafile is already opened */
+	public void open() throws IOException
+		{
+		if(isOpen()) return;
+		validateConfig();
+		
+		DataInputStream dis=new DataInputStream(new FileInputStream(getConfig().getSummaryFile()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dis*. The resource is referenced at lines 57, 58. The resource is closed at line 59. There are other execution paths that do not contain closure statements, for example, when DataInputStream.readLong throws an exception. Ensure *dis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).